qiskit/algorithms/minimum_eigen_solvers/adapt_vqe.py

```diff
@@ -0,0 +1,337 @@
+            )
+        if self.solver.quantum_instance is None:
+            raise AlgorithmError(
+                "A QuantumInstance or Backend must be supplied to run the quantum algorithm."
+            )
+        # We construct the ansatz once to be able to extract the full set of excitation operators.
+        self._tmp_ansatz._build()
+
+        prev_op_indices: List[int] = []
+        theta: List[float] = []
+        max_grad: Tuple[float, Optional[PauliSumOp]] = (0.0, None)
+        self._excitation_list = []
+        iteration = 0
+        while self._max_iterations is None or iteration < self._max_iterations:
+            iteration += 1
+            logger.info("--- Iteration #%s ---", str(iteration))
+            # compute gradients
+            cur_grads, expectation = self._compute_gradients(theta, operator)
+            # pick maximum gradient
+            max_grad_index, max_grad = max(
+                enumerate(cur_grads), key=lambda item: np.abs(item[1][0])
+            )
+            # store maximum gradient's index for cycle detection
+            prev_op_indices.append(max_grad_index)
+            # log gradients
+            if np.abs(max_grad[0]) < self._threshold:
+                if iteration == 1:
+                    raise QiskitError(
+                        "Gradient choice is not suited as it leads to all zero gradients gradients. "
+                        "Try a different gradient method."
+                    )
+                logger.info(
+                    "AdaptVQE terminated successfully with a final maximum gradient: %s",
+                    str(np.abs(max_grad[0])),
+                )
+                finishing_criterion = FinishingCriterion.CONVERGED
+                break
+            # check indices of picked gradients for cycles
+            if self._check_cyclicity(prev_op_indices):
+                logger.info("Final maximum gradient: %s", str(np.abs(max_grad[0])))
+                finishing_criterion = FinishingCriterion.CYCLICITY
+                break
+            # add new excitation to self._ansatz
+            self._excitation_list.append(max_grad[1])
+            theta.append(0.0)
+            # run VQE on current Ansatz
+            self._tmp_ansatz.operators = self._excitation_list
+            self.solver.ansatz = self._tmp_ansatz
+            self.solver.initial_point = theta
+            raw_vqe_result = self.solver.compute_minimum_eigenvalue(operator)
+            theta = raw_vqe_result.optimal_point.tolist()
+        else:
+            # reached maximum number of iterations
+            finishing_criterion = FinishingCriterion.MAXIMUM
+            logger.info("Maximum number of iterations reached. Finishing.")
+            logger.info("Final maximum gradient: %s", str(np.abs(max_grad[0])))
+
+        if finishing_criterion is False:
+            finishing_criterion = FinishingCriterion.UNKNOWN
```

Comment:
Is there any reason to have `FinishingCriterion.UNKNOWN` since the algorithm will raise an error setting the local variable there will never get seen right. I would argue that this a sanity check, an internal logic error if it gets to that point, ie breaks the loop without a condition. The is False test is really checking whether finishing_criterion was ever defined right since it never initialized/created anywhere before going into the loop. Arguably since the finishing criteria if part of result the unit testing out to check the set of termination paths that would arguably render this sanity check unnecessary - though I am not entirely opposed to keeping the error check check, but the FinishingCriterion.UNKNOWN seems like it ought to go